argoverse/evaluation/eval_detection.py

```diff
@@ -0,0 +1,402 @@
+
+        init_data = {dt_cls: MEASURE_DEFAULT_VALUES for dt_cls in self.detection_cfg.detection_classes}
+        summary = pd.DataFrame.from_dict(init_data, orient="index", columns=STATISTIC_NAMES)
+        summary_update = pd.DataFrame.from_dict(
+            self.summarize(data, cls_to_ninst), orient="index", columns=STATISTIC_NAMES
+        )
+
+        summary.update(summary_update)
+        summary = summary.round(SIGNIFICANT_DIGITS)
+        summary.index = summary.index.str.title()
+
+        summary.loc["Average Metrics"] = summary.mean().round(SIGNIFICANT_DIGITS)
+        return summary
+
+    def accumulate(self, gt_fpath: Path) -> Tuple[DefaultDict[str, np.ndarray], DefaultDict[str, int]]:
+        """Accumulate the true/false positives (boolean flags) and true positive errors for each class.
+
+        Args:
+            gt_fpath: Ground truth file path.
+
+        Returns:
+            cls_to_accum: Class to accumulated statistics dictionary of shape |C| -> (N, K + S) where C
+                is the number of detection classes, K is the number of true positive thresholds used for
+                AP computation, and S is the number of true positive errors.
+            cls_to_ninst: Mapping of shape |C| -> (1,) the class names to the number of instances in the ground
+                truth dataset.
+        """
+        log_id = gt_fpath.parents[1].stem
+        logger.info(f"log_id = {log_id}")
+        ts = gt_fpath.stem.split("_")[-1]
+
+        dt_fpath = self.detection_fpath / f"{log_id}/per_sweep_annotations_amodal/" f"tracked_object_labels_{ts}.json"
+
+        dts = np.array(read_label(str(dt_fpath)))
+        gts = np.array(read_label(str(gt_fpath)))
+
+        cls_to_accum = defaultdict(list)
+        cls_to_ninst = defaultdict(int)
+        for class_name in self.detection_cfg.detection_classes:
+            dt_filtered = filter_instances(
+                dts,
+                class_name,
+                filter_metric=self.detection_cfg.detection_metric,
+                max_detection_range=self.detection_cfg.max_detection_range,
+            )
+            gt_filtered = filter_instances(
+                gts,
+                class_name,
+                filter_metric=self.detection_cfg.detection_metric,
+                max_detection_range=self.detection_cfg.max_detection_range,
+            )
+
+            logger.info(f"{dt_filtered.shape[0]} detections")
+            logger.info(f"{gt_filtered.shape[0]} ground truth")
+            if dt_filtered.shape[0] > 0:
+                ranked_detections, scores = rank(dt_filtered)
+                metrics = self.assign(ranked_detections, gt_filtered)
+                cls_to_accum[class_name] = np.hstack((metrics, scores))
+
+            cls_to_ninst[class_name] = gt_filtered.shape[0]
+        return cls_to_accum, cls_to_ninst
+
+    def assign(self, dts: np.ndarray, gts: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
```

Comment:
as discussed with @benjaminrwilson  -- prefer for this function to be written in functional programming style and go into the utils, so we can add a simple unit test for it (to make sure overlapping detections with different confidence scores are accounted for properly)